src/v/cluster/data_migration_backend.h

```diff
@@ -89,11 +90,22 @@ class backend {
         errc ec;
     };
     class topic_scoped_work_state {
+        using self = topic_scoped_work_state;
         ss::abort_source _as;
+        retry_chain_node _rcn;
+        ss::shared_promise<errc> _promise;
```

Comment:
why is this a shared_promise? it's not often needed, and its use has been error prone in the past.

Reply:
Because I'm not sure it is always awaited only by one fiber.

In extreme cases when leadership changes a lot the following situation may happen:
1. We are a raft0 leader and the migration is in preparing state. We spawn a fiber to work towards prepared state and make an entry about it.
2. We lose leadership, and some other node, as a coordinator, transitioned the migration into executing state.
3. We regain raft0 leadership, and the migration is in executing state. We attempt to spawn a fiber to work towards prepared state and make an entry about it, but the entry is already there, so we abort the future and wait for it in order to guard from conflicting concurrent works on the same topic.
4. Yet again we lose leadership, and some other node, as a coordinator, completes the migration.
5. We regain raft0 leadership, the migration is gone, so we remove all the structures about it. In particular, we abort and wait for all related work to stop. It may happen that the original work from (1) still has not stopped yet.

I thought using a mutex (or a gate) and a variable holding a result instead, but felt like reinventing the wheel.